Extend ECDH interface

yanesca opened the pull request:
## Description
Extend ECDH interface to enable alternative implementations. The changes are not present in the default configuration.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
We want to enable alternative implementations for ECDH.

YES

yanesca commented:
@gilles-peskine-arm Could you please review this PR for design?

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -49,11 +52,38 @@ typedef enum
+    mbedtls_ecp_point Q;     /*!< The public key. */
+    mbedtls_ecp_point Qp;    /*!< The value of the public key of the peer. */
+    mbedtls_mpi z;           /*!< The shared secret. */
+    mbedtls_ecp_point Vi;    /*!< The blinding value. */
```
We're still not using `Vi`, `Vf` and `_d`, so how about removing them?

By the way: #1962

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +93,15 @@ typedef struct mbedtls_ecdh_context
     mbedtls_ecp_point Vf;    /*!< The unblinding value. */
     mbedtls_mpi _d;          /*!< The previous \p d. */
+#else
+    int point_format;        /*!< The format of point export in TLS messages. */
```
In new data structures, please refrain from using `int`. Use a suitable type instead: `size_t` for sizes, an unsigned type for most integers, an `enum` when only a small number of cases are possible... If I'm not mistaken, the point format is a 16-bit value, so the correct type would be `uint_least16_t` or, until we fully update our build chains to C99, `unsigned short`.

yanesca replied:
We are already using `uint16_t` in both the ECP and the SSL modules, therefore I think it would be safe to use (at least it wouldn't make things worse). Theoretically `uint_least16_t` would be nice, but it is not the most frequently used feature, are you sure it is worth the extra trust we need to put in our toolchains?

yanesca replied:
(It turned out that it is an 8-bit value, but we have `uint8_t` too in the SSL module already so the same argument holds. Also it is much less likely to have a platform without `uint8_t` than one without `uint16_t`)

mpg replied:
I think `stdint.h` is already allowed (it's the only C99 feature that's allowed so far). It's currently included in 26 places in the library, so I really don't see a reason to refrain from including it here.

Also, since we require `CHAR_BIT == 8`, I don't really see a reason to bother with the `uint_leastXX_t` types rather than using `uintXX_t` directly, as in practice all platforms that have the former and 8-bit bytes are going to have the later too. (Also, `grep 'int[0-9][0-9]*_t' library/*.c include/mbedtls/*.h | wc -l` prints `580` so if those types were a problem of our users, we would already know.)

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +93,15 @@ typedef struct mbedtls_ecdh_context
+    union
+    {
+        mbedtls_ecdh_context_mbed   *mbed_ecdh;
+    } ctx;                   /*!< Implementation specific context. */
```
Spellling: “implementation-specific context”.

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +93,15 @@ typedef struct mbedtls_ecdh_context
+#else
+    int point_format;        /*!< The format of point export in TLS messages. */
+    mbedtls_ecp_group_id grp;/*!< The elliptic curve used. */
+    mbedtls_ecdh_variant var;/*!< The ECDH implementation/structure used. */
```
Please state explicitly that `var` indicates which field of `ctx` is used, either in the description of `var` or in the description of `ctx`.

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -127,6 +166,16 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
 void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx );
 
+/**
+ * \brief           This function sets up the ECDH context.
```
Please add a paragraph explaining what this means and how this fits in the lifecycle of an ECDH context object (e.g. should I call `init` first?).

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -80,6 +80,7 @@
  * CHACHA20  3                  0x0051-0x0055
  * POLY1305  3                  0x0057-0x005B
  * CHACHAPOLY 2 0x0054-0x0056
+ * ECDH      1                  0x001B-0x001D
```
ECDH is above ECP and ECP is a high-level module, so ECDH should be a high-level module.

I'd prefer to refrain from adding a new module and declare all the errors in ECP.

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +93,15 @@ typedef struct mbedtls_ecdh_context
+    mbedtls_ecdh_variant var;/*!< The ECDH implementation/structure used. */
+    union
+    {
+        mbedtls_ecdh_context_mbed   *mbed_ecdh;
```
Why is this a pointer? Do we expect the alternative contexts to have wildly differing sizes? A pointer adds memory management overhead, which impacts RAM usage, performance (especially wrt caches) code size and complexity (this is visible in `mbedtls_ecdh_setup_internal`). For RAM alone, how much overhead there is per block depends wildly on the malloc implementation, but it's hard to go below 8 bytes of overhead, plus the pointer makes 12 bytes, so it's a net loss unless the size difference between two context types can be above 12 bytes.

wintersteiger replied:
From our end, both are fine. I will run some experiments to find out whether it makes a difference in our applications or not, but I don't expect a big difference.

yanesca replied:
The context sizes are 64 vs 300 where the latter can be reduced to 216. That seems acceptable so based on this and your comments I'll remove the pointer.

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -94,61 +101,195 @@ void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx )
+#else
+    mbedtls_ecdh_context_mbed *real_ctx;
+
+    if( ctx->ctx.mbed_ecdh != NULL )
```
I think `mbed_ecdh` should be the structure, not a pointer, which would simplify this function.

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -94,61 +101,195 @@ void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx )
+    ctx->grp = MBEDTLS_ECP_DP_NONE;
+#endif
+
+    mbedtls_ecp_group_free( &real_ctx->grp );
```
I think it would be better to put the code that handles the “MBED” variant context in a separate function, to pave the way for a build that doesn't include this context variant. Make `mbedtls_ecdh_free_internal` take a `mbedtls_ecdh_context_mbed*` as argument. This goes for other functions with a similar structure.

yanesca replied:
Separating them would result in duplicating a lot of code and this means extra maintenance and opportunity for errors. I think that removing the legacy version is pretty straightforward and therefore I would prefer to keep them as they are.

gilles-peskine-arm commented on library/ssl_cli.c:
```diff
@@ -2021,14 +2027,20 @@ static int ssl_check_server_ecdh_params( const mbedtls_ssl_context *ssl )
 #endif
         return( -1 );
 
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
```
This 7-line pattern (`#if LEGACY_CONTEXT`, `DEBUG`, `#ELSE`, `if`, `DEBUG`, continuation, `#endif`) is repeated many times and is due to evolve to support other context variants. It should be an auxiliary function or macro.

gilles-peskine-arm commented on library/ssl_srv.c:
```diff
@@ -3901,7 +3920,14 @@ static int ssl_parse_client_key_exchange( mbedtls_ssl_context *ssl )
+        if( ssl->handshake->ecdh_ctx.var == MBEDTLS_ECDH_VARIANT_MBED )
+            MBEDTLS_SSL_DEBUG_ECP( 3, "ECDH: Qp",
+                                   &ssl->handshake->ecdh_ctx.ctx.mbed_ecdh
+                                    ->Qp );
```
That's a pretty awkward line break. I'd prefer a 81-character line.

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -187,7 +199,7 @@ void mbedtls_ecdh_free( mbedtls_ecdh_context *ctx );
  *                  ServerKeyExchange payload.
  *
- *                  This is the first function used by a TLS server for ECDHE
+ *                  This is the second function used by a TLS server for ECDHE
```
Please add a reference to the first function or to a place where all the successive functions are listed.

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -280,7 +292,7 @@ int mbedtls_ecdh_make_public( mbedtls_ecdh_context *ctx, size_t *olen,
  *              payload.
  *
- *              This is the second function used by a TLS server for ECDH(E)
+ *              This is the third function used by a TLS server for ECDH(E)
```
Please add a reference to the first two functions or to a place where all the successive functions are listed.

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -386,6 +386,20 @@
+ *              MBEDTLS_ECDH_COMPUTE_SHARED_ALT for curves which have formally
+ *              verified implementation available.
+ */
+#define MBEDTLS_ECDH_LEGACY_CONTEXT
```
Why are we introducing yet another preprocessor symbol? Can we just not deprecate the old interface using the existing methods?

yanesca replied:
Also this is just temporary, this flag would be removed in the next major version.

Another benefit is that until that time we have the freedom to further improve the context and do so in several iterations. With the standard deprecation method this wouldn't be feasible.

We will be adding another primitives and working on other API improvements and I think that the opportunity to use the lessons learned there or just to provide consistency is very well worth temporarily adding a new preprocessor symbol.

sbutcher-arm replied:
So this is sort of the opposite of a deprecated interface, it's an interface we don't want on by default yet, but we want to be present alongside the current interface. Historically we've released evaluation code on different branches, not like this, so there's no precedent, which is I think the correct answer to my question.

Therefore we're transitioning from one API to another, and want to retain one while we move to the other. In which case it needs to be clear that the older one will be phased out. It also needs to be clear that this isn't just an 'option' but a stepping stone to changing the API.

Right now, if a user were to look through `config.h` and saw this, it looks like yet another optional feature like all the others. So does it need to be in `config.h`? You can remove the symbol from `config.h` and it'll work just as well, but won't confuse users.

I take it that we want to have the new interface off by default for the moment, and we want to test with the existing the interface, so what advantage are we providing by exposing this to users?

Patater replied:
"it needs to be clear that the older one will be phased out." -- Should we deprecate the current interface, yet still use the deprecated interface by default?

The argument for having the option in `config.h` seems to be "because all configuration options are in config.h," regardless of whether those options are useful to users. I'm not saying this is a strong argument.

gilles-peskine-arm replied:
The old interface is deprecated. We want people to use the new interface, but we don't want to break the API. Unfortunately we haven't found a way to get the benefits of the new implementation without changing the context structure.

Our intended use case is to add EC implementations that are better at least in some respects than the default implementation, but require a different layout of the context structure. The new context structure allows EC implementations to be added without breaking the API.

We could do a better job of documenting this, but I don't see a better way to encode it.

We do have precedents. For example `MBEDTLS_CMAC_C` defaults off. Not because we don't want people to use CMAC, but because it changes `mbedtls_cipher_context_t`.

sbutcher-arm replied:
> The argument for having the option in config.h seems to be "because all configuration options are in config.h," regardless of whether those options are useful to users.

@Patater - That's false. There are examples of configurable options which are local and aren't in `config.h` but are configurable. `config.h` is the public interface of configurable options. There are other options elsewhere such as `MBEDTLS_HAVE_INT32` which aren't in `config.h`.

> The old interface is deprecated. We want people to use the new interface, but we don't want to break the API. 

If the old interface is deprecated, it should get a deprecated warning on build, aside from the issues of how we enable this new interface.

yanesca commented:
@gilles-peskine-arm - Thank you for your feedback, I think I have addressed all of your remarks.

@sbutcher-arm - Both @gilles-peskine-arm and I have left replies to your comment on the config.h flag. Could you please approve if you can accept our arguments and reply if you don't?

I haven't removed the needs work label yet, because I would like to rebase after I have the design approval from both of you to facilitate further reviews.

yanesca commented:
@sbutcher-arm I have removed the flag from `config.h`, could you please have a look at the last commit and approve the PR if you give your design approval or comment if you don't?

gilles-peskine-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -36,6 +36,15 @@
+ * must contain a warning stating that the feature breaks backward
+ * compatibility.
+ */
+#define MBEDTLS_ECDH_LEGACY_CONTEXT
```
This makes it impossible to switch off the legacy context without editing the file.

This flag belongs in `config.h`.

yanesca replied:
This flag is not a standalone feature at this point. The new context will be used by other features, like the upcoming alternative implementation. Those are features that need a flag in `config.h` and need to conditionally turn off the legacy flag here to make use of the new context.

gilles-peskine-arm replied:
But in the meantime, we can't even test a build with `MBEDTLS_ECDH_LEGACY_CONTEXT` not defined.

yanesca replied:
Yes, it can only be tested manually as it is. This is a problem, but only until the first feature that uses this option. After that it will be tested whenever the corresponding feature is tested.

If this is not enough, we can extend `config.pl` with new commands (eg setmod/unsetmod) to set/unset flags in module headers to enable automatic testing.

wintersteiger replied:
Would it help if we merged our changes with this PR so that there is at least one feature that uses the new macros? Also, I think it would be possible to base our PR on the branch of this PR, so that they can be merged in tandem.

yanesca replied:
I feel that it would be cleaner to handle the two PRs separately for now.

sbutcher-arm replied:
No changes are necessary to us `config.pl` on this. You can just do this:
```
./scripts/config.pl unset MBEDTLS_ECDH_LEGACY_CONTEXT --file include/mbedtls/ecdh.h
```
That will comment out the configuration.

gilles-peskine-arm commented:
I'm ok with leaving a testing gap if we record what's been tested manually and we have a clear plan (tracked in an issue, please) to fix this.

yanesca commented:
@gilles-peskine-arm I have documented manual testing and suggested a plan to fix in IOTCRYPT-433.

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -102,53 +181,103 @@ void mbedtls_ecdh_free( mbedtls_ecdh_context *ctx )
     int ret;
     size_t grp_len, pt_len;
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_context *real_ctx = ctx;
```
I'd prefer to change the naming: `ctx_arg` and `ctx` instead of `ctx` and `real_ctx`. This is only a suggestion, I don't insist on it, as long as it's consistent between functions.

yanesca replied:
I wasn't happy with the `real_ctx` either, but couldn't come up with anything better.

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -91,7 +98,79 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+#endif
+}
+
+static void mbedtls_ecdh_free_internal( mbedtls_ecdh_context *ctx )
```
How about changing the `xxx_internal` functions as follows?
```
#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
typedef mbedtls_ecdh_context mbedtls_ecdh_context_mbed;
#endif
// …
static int xxx_internal( mbedtls_ecdh_context_mbed* ctx )
{ … }
int xxx( mbedtls_ecdh_context *ctx )
{
#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
    return( xxx_internal( ctx ) );
#else
    switch( ctx->var ) {
        case MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0:
            return( xxx_internal( &ctx->ctx.mbed_ecdh ) );
    }
#endif
}
```

This simplifies the internal functions without making the public wrappers more complex.

yanesca replied:
I think it is a brilliant idea!

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -240,29 +444,80 @@ int mbedtls_ecdh_read_public( mbedtls_ecdh_context *ctx,
+            return( mbedtls_ecdh_calc_secret_internal( ctx, olen, buf, blen,
+                                                       f_rng, p_rng ) );
+        default:
+            return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
```
If you're changing or adding a line with `return` without parentheses, please add parentheses as our house style requires.

sbutcher-arm commented on include/mbedtls/debug.h:
```diff
@@ -221,6 +227,36 @@ void mbedtls_debug_print_crt( const mbedtls_ssl_context *ssl, int level,
+ * \param file      file the error has occurred in
+ * \param line      line number the error has occurred in
+ * \param ecdh      the ECDH context
+ * \param attr  the identifier of the attribute being output
```
Some alignment problem here.

sbutcher-arm commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +102,16 @@ typedef struct mbedtls_ecdh_context
     mbedtls_ecp_point Vf;    /*!< The unblinding value. */
     mbedtls_mpi _d;          /*!< The previous \p d. */
+#else
+    uint8_t point_format;        /*!< The format of point export in TLS messages. */
```
Alignment.

Patater commented on include/mbedtls/ecp.h:
```diff
@@ -538,19 +539,39 @@ int mbedtls_ecp_group_load( mbedtls_ecp_group *grp, mbedtls_ecp_group_id id );
+ *
+ * \return          \c 0 on success.
+ * \return          #MBEDTLS_ERR_ECP_BAD_INPUT_DATA if input is invalid.
+ * \return          #MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE if the group is not
```
How can `mbedtls_ecp_tls_read_group_id()` return `MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE`? It doesn't seem possible with the current implementation, but perhaps there is a use case for it in some future implementation?

yanesca replied:
It is possible with the current implementation if:
- The format is compressed or malformed or
- The curve in question is not enabled in `config.h` or not recognised at all

Patater replied:
OK, I see this from the following code in that function that I didn't notice before:

```C
    if( ( curve_info = mbedtls_ecp_curve_info_from_tls_id( tls_id ) ) == NULL )
        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
```

mpg commented on library/debug.c:
```diff
@@ -365,4 +365,54 @@ void mbedtls_debug_print_crt( const mbedtls_ssl_context *ssl, int level,
+#else
+    switch( ecdh->var )
+    {
+        default:
```
Question: how are implementors of alternative implementations supposed to provide their own version of the debug callback? I'm probably missing something but currently it looks like they have to edit this `switch` statement...

yanesca replied:
Yes, they do have to edit this switch statement. All of the corresponding functions have this structure and they need to edit all of the `switch` statements that relay the function call.

mpg commented on include/mbedtls/ecdh.h:
```diff
@@ -63,6 +106,17 @@ typedef struct mbedtls_ecdh_context
+#else
+    uint8_t point_format;    /*!< The format of point export in TLS messages as
+                                  defined in RFC 4492. */
+    mbedtls_ecp_group_id grp;/*!< The elliptic curve used. */
```
Minor: I'd prefer `gid` or something else than `grp` here to avoid confusion. Usually `grp` means an `mbedtls_ecp_group`.

Patater commented on library/ecdh.c:
```diff
@@ -91,64 +102,161 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+
+    mbedtls_ecp_group_init( &ctx->grp );
+    mbedtls_ecp_point_init( &ctx->Q   );
+    mbedtls_ecp_point_init( &ctx->Qp  );
```
We were previously just memsetting the context to 0 for the legacy context. Now, we memset the context to 0 and also call these init functions, while with the non-legacy context we simply memset to 0. Is this the right way around? Shouldn't the legacy context configuration behave identically to before, and the new context configuration call the individual init functions?

yanesca replied:
It doesn't make any difference, because currently `mbedtls_ecp_group_init()` just calls memset, and `mbedtls_ecp_point_init()` has an equivalent behaviour to memset (initialising the coordinates to `+0` versus `0`).

But you are right, it is better to be more robust and call the `_init()` functions.

Patater replied:
After your latest changes, we appear to be calling the init functions for both the legacy configuration and the non-legacy configuration.

I was asking if we actually needed to call the `mbedtls_ecp_group_init()` and `mbedtls_ecp_point_init()` functions for the legacy configuration, as we didn't before this PR. This PR added them for the legacy configuration.

yanesca replied:
I don't remember why I added it, I think it is better to not change behaviour, I will restore both cases to the old behaviour.

Patater replied:
I don't think its necessary to the legacy configuration to call the init functions (they weren't called before), but should we call the init functions for the new non-legacy configuration?

Patater commented on library/ecdh.c:
```diff
@@ -91,64 +102,161 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
-    mbedtls_ecp_point_free( &ctx->Vf  );
     mbedtls_mpi_free( &ctx->d  );
     mbedtls_mpi_free( &ctx->z  );
-    mbedtls_mpi_free( &ctx->_d );
```
The legacy context still contains `Vi`, `Vf`, and `_d`. Shouldn't we still free these for the legacy context configuration case? (Perhaps free these in `mbedtls_ecdh_free()` before the call to `mbedtls_ecdh_free_internal()` in the legacy configuration case.)

Patater commented on include/mbedtls/debug.h:
```diff
@@ -221,6 +227,36 @@ void mbedtls_debug_print_crt( const mbedtls_ssl_context *ssl, int level,
+ * \attention       This function is intended for INTERNAL usage within the
+ *                  library only.
+ */
+void mbedtls_debug_printf_ecdh( const mbedtls_ssl_context *ssl, int level,
```
Do we imply that all `debug` functions are internal? Is this function somehow different than other debug functions? Should it have a name ending in `_internal`?

yanesca replied:
The only `debug` function that is supposed to be called by user is `mbedtls_debug_set_threshold()`.  The others are all marked internal in the documentation. This is not different than the other debug functions. Other debug functions that are marked as internal do not have an `_internal` postfix therefore I think the naming is more consistent like this.

Patater replied:
OK. Agreed.

Patater commented on include/mbedtls/ecdh.h:
```diff
@@ -127,6 +181,24 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+ *
+ * \return          \c 0 on success.
+ */
+int mbedtls_ecdh_setup( mbedtls_ecdh_context *ctx, mbedtls_ecp_group_id grp );
```
Old applications aren't going to call `mbedtls_ecdh_setup()`. How will ECDH still work for old applications that aren't changing their code? Is it unnecessary to call `mbedtls_ecdh_setup()` when using the legacy context configuration? If it is now necessary, this is an API break in the default configuration, which we potentially care to avoid.

yanesca replied:
Old applications had to do by hand all the work that `mbedtls_ecdh_setup()` does. In the legacy context configuration it works either way. Old applications do not need to change their code, new applications may use `mbedtls_ecdh_setup()` or set it by hand like the old ones.

Patater replied:
OK

Patater commented on library/ssl_cli.c:
```diff
@@ -2021,14 +2027,15 @@ static int ssl_check_server_ecdh_params( const mbedtls_ssl_context *ssl )
 
-    MBEDTLS_SSL_DEBUG_ECP( 3, "ECDH: Qp", &ssl->handshake->ecdh_ctx.Qp );
+    MBEDTLS_SSL_DEBUG_ECDH( 3, &ssl->handshake->ecdh_ctx,
+                            MBEDTLS_DEBUG_ECDH_QP );
```
Do we have any tests that depend on the format of these debug messages? If the output of the debug message changes and a test depends on it, we might find some tests start failing.

yanesca replied:
We do have tests that depend on the format of these debug messages, but as far as I can tell this does not change the output.

(Also the CI has passed on both branches)

Patater commented on library/ssl_srv.c:
```diff
@@ -3039,8 +3039,8 @@ static int ssl_prepare_server_key_exchange( mbedtls_ssl_context *ssl,
-        if( ( ret = mbedtls_ecp_group_load( &ssl->handshake->ecdh_ctx.grp,
-                                       (*curve)->grp_id ) ) != 0 )
+        if( ( ret = mbedtls_ecdh_setup( &ssl->handshake->ecdh_ctx,
+                                        (*curve)->grp_id ) ) != 0 )
```
Why did we need to change this to `mbedtls_ecdh_setup()`? Couldn't we keep calling `mbedtls_ecp_group_load()`? If not, we've broken the API in the default configuration.

yanesca replied:
Because the new context the `grp` member has different type and this call wouldn't compile in the non-legacy case. It would still work in the default configuration, but we want this to work in both. I don't think we have broken the API in the default configuration with this.

Patater replied:
OK, so we only changed the API such that if you want to work with the new non-legacy option, you need to change to call `mbedtls_ecdh_setup()`. This is fine.

Patater commented on tests/suites/test_suite_ecdh.function:
```diff
@@ -131,12 +131,14 @@ void ecdh_exchange( int id )
     memset( &rnd_info, 0x00, sizeof( rnd_pseudo_info ) );
 
-    TEST_ASSERT( mbedtls_ecp_group_load( &srv.grp, id ) == 0 );
+    TEST_ASSERT( mbedtls_ecdh_setup( &srv, id ) == 0 );
```
Should we keep a test with `mbedtls_ecp_group_load()` as well to ensure we don't break the API?

yanesca replied:
We are using `mbedtls_ecp_group_load()` in the other two test functions in this suite, I think that should be enough to ensure that we don't break the API.

Patater replied:
OK

Patater replied:
Follow-up questions: will those two other tests still work when the new configuration is used with an alternate implementation of a particular group? In general, which of our tests will be available for alternative group implementers to check their work with, and how many depend on the group being Mbed TLS implemented (i.e. using `MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0`)?

yanesca replied:
Ok, I think I misread the tests, the other two tests don't use these contexts, they exercise a lower level API.

I'll add a test to exercise the legacy workflow.

Patater commented on tests/suites/test_suite_ecdh.function:
```diff
@@ -150,8 +152,10 @@ void ecdh_exchange( int id )
+    TEST_ASSERT( mbedtls_ecdh_calc_secret( &cli, &res_len, res_buf, 1000,
+                                   NULL, NULL ) == 0 );
+    TEST_ASSERT( len == res_len );
+    TEST_ASSERT( memcmp( buf, res_buf, len ) == 0 );
```
Did we have to change these asserts (no longer using `mbedtls_mpi_cmp_mpi()` because MPI might not be used in the underlying implementation? Did we change the format of data returned by `mbedtls_ecdh_calc_secret()` (whether in the default configuration or otherwise)?

yanesca replied:
Yes, that was the reason. No, we didn't, it is unsigned fixed size integer in a big endian byte order like it was before.

Patater replied:
OK. So, again, the old way works with the legacy configuration still, and we've only changed this test to use the new way required so that it can use both the old and new configuration.

mpg commented:
Note: this is going to have fairly non-trivial conflicts with https://github.com/ARMmbed/mbedtls-restricted/pull/390 - so when on of the PRs is merged, the other will need re-work and ideally re-review by someone involve in the other PR to ensure conflicts are resolved in a way that doesn't break any of them.

yanesca commented:
I wasn't planning on writing a separate documentation, I thought we could just place comments the code, but I am happy to write one if the comments are insufficient or would grow unreasonably big.

To provide an alternative the integrator has to:
1. Add their own constant to `mbedtls_ecdh_variant`
2. Add their own structure to the `ctx` union in `mbedtls_ecdh_context`
3. Implement their functions: `mbedtls_debug_printf_ecdh_xxx()`, `mbedtls_ecdh_setup_xxx()`, `mbedtls_ecdh_free_xxx()`, `mbedtls_ecdh_make_params_xxx()`, `mbedtls_ecdh_read_params_xxx()`, `mbedtls_ecdh_get_params_xxx()`, `mbedtls_ecdh_make_public_xxx()`, `mbedtls_ecdh_read_public_xxx()`, `mbedtls_ecdh_calc_secret_xxx()`
4. Edit the `switch` statements in the corresponding functions to dispatch to their variants.

This seems too long to write in a comment and could be extended to a Knowledge Base article, but given the temporary nature I don't think if it would be appropriate.

yanesca commented:
@mpg That is bad news. I think we could make the two options mutually exclusive and that should somewhat simplify resolving the conflicts.

yanesca commented:
@gilles-peskine-arm , @Patater Thank you for your reviews!

I have answered to your comments and pushed new commits to address them. The PR is ready for your review again.

Patater commented on library/ecdh.c:
```diff
@@ -144,9 +136,6 @@ int mbedtls_ecdh_setup( mbedtls_ecdh_context *ctx, mbedtls_ecp_group_id gid )
             ctx->gid = gid;
-            mbedtls_ecp_group_init( &ctx->ctx.mbed_ecdh.grp );
-            mbedtls_ecp_point_init( &ctx->ctx.mbed_ecdh.Q   );
-            mbedtls_ecp_point_init( &ctx->ctx.mbed_ecdh.Qp  );
```
This is for the new configuration and we might want to keep this in. Needs thinking about.

yanesca replied:
I think it is simpler to not call these because
- this makes the initialisation more similar and less error prone
- this is how we did it in the past
- IIRC PSA mandates that `memset()` -ing to 0 is a valid initialisation of contexts so this will be a valid way of initialisation in the future too

Patater commented:
retest

Patater commented:
CI failure is due to FreeBSD timing test failing

wintersteiger commented:
Just checking - is it just the conflict in `ChangeLog` that's waiting to be resolved before this will be merged?

yanesca commented:
I have rebased on current head, here is the old history:
https://github.com/yanesca/mbedtls/tree/kex-into-pk-v5-pre-rebase

I have cleaned up the history before rebasing, here is the result:
https://github.com/yanesca/mbedtls/tree/kex-into-pk-v5-old-head

@gilles-peskine-arm @Patater Can you please give it another round of review?

yanesca commented:
I forgot to mention that I have noticed and fixed 3 issues while rebasing:
- A minor style issue in the documentation
- Some macros and an include in ecdh.c that weren't used anymore                 
- Internal functions had `mbedtls_` prefix and this is against Mbed TLS coding standards

RonEld commented on include/mbedtls/ecdh.h:
```diff
@@ -70,7 +116,26 @@ typedef struct mbedtls_ecdh_context
+    union
+    {
+        mbedtls_ecdh_context_mbed   mbed_ecdh;
+    } ctx;                      /*!< Implementation-specific context. The
```
The comment here is confusing. Also, this is a union of what? I only see `mbed_ecdh`

RonEld replied:
I have just seen that PR. This means that every alternative implementation should modify `ecdh.h` to add their context?
I don't agree with this design. I think it is better to have some `_ALT` context , that every alternative implementation should define in their own specific `alt` header file .
I understand this is a bit late in the design process, but I strongly suggest rethinking this.

gilles-peskine-arm replied:
The goal of this mechanism different from the ALT one. Here we expect a small number of alternatives that go through us, not independent overriding of software implementations by a hardware driver. And we want to be able to link multiple alternatives.

RonEld commented on include/mbedtls/ecdh.h:
```diff
@@ -70,7 +116,26 @@ typedef struct mbedtls_ecdh_context
+    uint8_t restart_enabled;    /*!< The flag for restartable mode. Functions of
+                                  an alternative implementation not supporting
+                                  restartable mode must return
+                                  MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE error if
```
The error should be `MBEDTLS_ERR_PLATFROM_FEATURE_UNSUPPORTED` once #2054 will be merged

RonEld replied:
Also, in #2165 we agreed that an alternative implementation and the restartable feature should not co exist, as they don't make sense to be together. We also added protection for that in `check_config.h`
We should consider this in the design of this PR

yanesca replied:
About the error: that is a good point, this needs to be fixed in whichever PR gets merged later. I suggest we make a note of this at #2054 too.

About the restartable feature: This PR aims for a tighter integration than #2165 and unlike there, here we have a sensible way of making them coexist. I think in some scenarios it makes perfect sense to use them together. The alternative implementation might be restartable too and even if it is not, it still might be so much faster that it does not need it. Of course a typical alternative implementation will not implement all of the curves Mbed TLS supports.

yanesca commented:
I have rebased the PR again to the current development branch. Also I have updated the documentation to require returning the new `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` error.

yanesca commented:
@gilles-peskine-arm, @Patater Could you please give it another round of review?

gilles-peskine-arm commented on library/ecdh.c:
```diff
@@ -278,53 +414,114 @@ int mbedtls_ecdh_get_params( mbedtls_ecdh_context *ctx, const mbedtls_ecp_keypai
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    if( ( ret = mbedtls_ecdh_setup( ctx, key->grp.id ) ) != 0 )
```
The code has changed from `mbedtls_ecp_group_copy( &ctx->grp, &key->grp )` to `mbedtls_ecp_group_load( &ctx->grp, key->grp.id )`. This is correct only if there is a single group with any given ID. This applies to our own code, but might some users have added their own groups? I think that mandating that groups have distinct IDs is a reasonable requirement, but was it already a requirement before for ECDH? Is it properly documented?

yanesca replied:
I have added clarification about this requirement.

Patater commented on include/mbedtls/ecp.h:
```diff
@@ -159,6 +159,10 @@ mbedtls_ecp_point;
  *
+ * \note        Alternative implementations must keep the group IDs distinct. If
+ *              two group structures have the same ID, then they must be the
+ *              identical.
```
"must be the identical" what? Maybe you don't want "the" here.

Patater commented on library/ecdh.c:
```diff
@@ -124,43 +128,85 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+            ctx->grp_id = grp_id;
+            ecdh_init_internal( &ctx->ctx.mbed_ecdh );
+            return( ecdh_setup_internal( &ctx->ctx.mbed_ecdh,
+                                                 grp_id ) );
```
Funny alignment.
